src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,310 @@
+     *
+     * @param stack the item stack to check
+     */
+    public boolean is(ItemStack stack) {
+        return ItemStack.isSameItemSameComponents(stack, innerStack);
+    }
+
+    /**
+     * {@return true if the item instance matches the backing instance value}
+     *
+     * @param item the item to check
+     */
+    public boolean is(ItemLike item) {
+        return is(item.asItem());
+    }
+
+    /**
+     * Tests an {@link ItemStack} predicate with the inner stack.
+     *
+     * @param predicate Predicate to perform the test with
+     * @return {@code true} if the test passed
+     */
+    public boolean test(Predicate<ItemStack> predicate) {
+        return predicate.test(innerStack);
+    }
+
+    @Override
+    public boolean isComponentsPatchEmpty() {
+        return innerStack.isComponentsPatchEmpty();
+    }
+
+    @Override
+    public ItemResource withPatch(DataComponentPatch patch) {
+        if (isEmpty() || patch.isEmpty() || innerStack.getComponentsPatch().equals(patch))
+            return this;
+
+        ItemStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+
+        return ItemResource.of(stack);
+    }
+
+    @Override
+    public <D> ItemResource with(DataComponentType<D> type, D data) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.set(type, data);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
+
+        return ItemResource.of(stack);
+    }
+
+    //This is overridden to return ItemResource to allow method chaining
+    @Override
+    public <D> ItemResource with(Supplier<? extends DataComponentType<D>> type, D data) {
+        return with(type.get(), data);
+    }
+
+    @Override
+    public ItemResource without(DataComponentType<?> type) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.remove(type);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
+
+        return ItemResource.of(stack);
+    }
+
+    //This is overridden to return ItemResource to allow method chaining
+    @Override
+    public ItemResource without(Supplier<? extends DataComponentType<?>> type) {
+        return without(type.get());
+    }
+
+    @Override
+    public DataComponentMap getComponents() {
+        return innerStack.immutableComponents();
+    }
+
+    @Override
+    public DataComponentPatch getComponentsPatch() {
+        return innerStack.getComponentsPatch();
+    }
+
+    /**
+     * Creates an {@link ItemStack} of the specified count.
+     *
+     * @param count The amount of the item the stack should have. Must be non-negative.
+     * @return A new copy of the inner item stack with the specified count.
```

Comment:
This is an implementation note. The fact that there is an inner stack is irrelevant to the API